### DISCUSSION TITLE
App crash at Startup

### DISCUSSION THREAD

**Discussion by LaurentInSeattle** [AvaloniaUI/Avalonia#17672] [Q&A] [answered]
Hi all, 

On Windows 10 (fully patched) After upgrading to .Net 9 and the latest Avalonia nugets at 11.2.2,  my app crashes at startup with the following trace: 

System.PlatformNotSupportedException: Operation is not supported on this platform.
   at Avalonia.Threading.Dispatcher.MainLoop(CancellationToken cancellationToken)
   at Avalonia.Controls.ApplicationLifetimes.ClassicDesktopStyleApplicationLifetime.StartCore(String[] args)
   at Avalonia.Controls.ApplicationLifetimes.ClassicDesktopStyleApplicationLifetime.Start(String[] args)
   at Avalonia.ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime(AppBuilder builder, String[] args, Action`1 lifetimeBuilder)
   at App.Desktop.Program.Main(String[] args) in C:..............Desktop\Program.cs:line 12

An unhandled exception of type 'System.PlatformNotSupportedException' occurred in Avalonia.Base.dll
Operation is not supported on this platform.

It seems like this area is causing the throw: 

	public void MainLoop(CancellationToken cancellationToken)
	{
		if (_controlledImpl == null)
		{
			throw new PlatformNotSupportedException();    // <== Here !
		}

I have "upgraded" the same way five other apps (using the same libraries) with no problem and only one is causing this issue.

Any idea ? 
Thanks!

**Comment by timunie** [comment]
Well... Try to narrow down the issue into a minimum sample. Then attach it for double checking.

**Comment by LaurentInSeattle** [comment, marked as answer]
Turns out that subscribing to the "ShutdownStarted" event in the way shown below is the cause of the issue.
(This was added to debug a possible failure at startup.)

class Program
{
    [STAThread]
    public static void Main(string[] args) 
        => BuildAvaloniaApp().StartWithClassicDesktopLifetime(args);

    public static AppBuilder BuildAvaloniaApp()
    {
        var builder = 
            AppBuilder.Configure<App>().UsePlatformDetect().WithInterFont().LogToTrace();
        Dispatcher.UIThread.ShutdownStarted += UIThread_ShutdownStarted;
        return builder;
    }

    private static void UIThread_ShutdownStarted(object? sender, EventArgs e)
    {
        if (Debugger.IsAttached) { Debugger.Break(); }
    } 
}

**Reply by kekekeks** [reply]
It's not safe to access dispatcher APIs before initialization is completed. Add a callback via builder.AfterSetup and use dispatcher there

**Comment by MisakaRehana** [comment]
same problem here
but what was so strange that I didn't access any dispatcher APIs here (cc. @kekekeks ) , and a plain blank main window still causes the problem.
avalonia 11.3.7 + .NET 8.0
totally strange, huh?

```
Unhandled exception. System.PlatformNotSupportedException: Operation is not supported on this platform.
   at Avalonia.Threading.Dispatcher.MainLoop(CancellationToken cancellationToken)
   at Avalonia.Controls.ApplicationLifetimes.ClassicDesktopStyleApplicationLifetime.StartCore(String[] args)
   at Avalonia.Controls.ApplicationLifetimes.ClassicDesktopStyleApplicationLifetime.Start(String[] args)
   at Avalonia.ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime(AppBuilder builder, String[] args, Action`1 lifetimeBuilder)
   at MyApp.Program.Main(String[] args)
   at MyApp.Program.<Main>(String[] args)
```

```csharp
public class MyAppType : Application
{
	public IClassicDesktopStyleApplicationLifetime DesktopAppLifetime => (IClassicDesktopStyleApplicationLifetime)ApplicationLifetime!;

	/// <summary>
	/// Same as WPF's <c>OnStartup</c>.
	/// </summary>
	public override void OnFrameworkInitializationCompleted()
	{
		try
		{
			if (Design.IsDesignMode) // design time will not implement ApplicationLifetime
			{
				return;
			}

			DesktopAppLifetime.ShutdownMode = ShutdownMode.OnExplicitShutdown;
			DesktopAppLifetime.MainWindow = null;

			var splash = new WinSplash();
			splash.Show();
		}
		finally
		{
			base.OnFrameworkInitializationCompleted();
		}
	}
```

yes, the splash window was shown, and then it crashed.

**Comment by MisakaRehana** [comment]
and [here](https://github.com/MisakaRehana/project-afs/tree/6d428511f4881022e5c98bc9f71e5e0880b67b9d) is a small project that can reproduce this issue

**Reply by LaurentInSeattle** [reply]
I took a quick look. It seems like your app is creating the avalonia builder twice. That could be a problem.

**Reply by MisakaRehana** [reply]
> I took a quick look. It seems like your app is creating the avalonia builder twice. That could be a problem.

Thanks for your hint, I will take a look at it
